Add LoRA Implementation

anwai98 opened the pull request.

anwai98 commented on micro_sam/training/util.py:
```diff
@@ -42,6 +42,7 @@ def get_trainable_sam_model(
     checkpoint_path: Optional[Union[str, os.PathLike]] = None,
     freeze: Optional[List[str]] = None,
     return_state: bool = False,
+    get_lora: bool = False,
```
Yes, it does make sense to expose this a bit. For simplicity, I have kept botht he function signatures `use_lora` and `rank` in the signatures. A modular choice would be to expose just the `rank` and decide on LoRA or full finetuning, but I would leave it to a simpler structure now for the experiments. Thanks for the mention.

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,155 @@
+class PEFTBase(nn.Module):
+    """PEFTBase is an interface to implement specific PEFT-based methods.
+    """
+    def __call__(
```
It would make more sense to have an `__init__` here, that defines a common signature for all PEFT adapter child classes.

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,155 @@
+class LoRASurgery(PEFTBase):
+    """Operates on the attention layers for performing low-rank adaptation.
+    """
+    def __call__(
```
Use `__init__` instead, set all the relevant weights as members (`w_a_linear` etc.).  If needed implement a `get_qkv` function that you can call in `PEFT_Sam` to do the block update.

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,155 @@
+        return block, [w_a_linear_q, w_b_linear_q, w_a_linear_v, w_b_linear_v]
+
+
+class _LoRA_Sam(nn.Module):
```
Absorb this into the `LoRASurgery` class.

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,155 @@
+
+        # create for storage, then we can init them or load weights
+        self.w_As = []  # These are linear layers
+        self.w_Bs = []
```
Create a list `self.peft_blocks = []` instead.

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,155 @@
+            w_a_linear_q, w_b_linear_q, w_a_linear_v, w_b_linear_v = linear_layers
+
+            self.w_As.extend([w_a_linear_q, w_a_linear_v])
+            self.w_Bs.extend([w_b_linear_q, w_b_linear_v])
```
Do this instead:
```python
peft_block_i = self._peft_module(...)
blk.attn.qkv  = peft_block.get_qkv()
self.peft_blocks.append(peft_block_i)
```

constantinpape replied:
We don't need to expose the weights `w_a_linear` etc. at all then, they live in the `peft_block`.

constantinpape replied:
On second thought: I think it's even better to do `blk.attn.qkv  = peft_block.get_qkv()` in the `__init__` of the `PEFTBlock`, by passing `blk` to it. This way the details of the adaptation are all abstracted away.

constantinpape reviewed (commented):
I think it would make more sense to design the PEFT adapters as proper pytorch modules, see the comments for details.

anwai98 commented:
@constantinpape Thanks for all the suggestions. I thought about all your feedback ,and imcorporated them now in a much more flexible design. Let me know if I missed something or you spot some more room for improvement. Thanks!

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,99 @@
+                range(len(model.image_encoder.blocks))
+            )
+
+        self.peft_module = peft_module
```
I think we need a list here that keeps the instantiated peft modules:
```
self.peft_blocks = []
```

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,99 @@
+            if t_layer_i not in self.peft_layers:
+                continue
+
+            blk.attn.qkv = self.peft_module(rank=rank, block=blk)
```
I would change this in two ways: Add the instantiated module to the list and move the qkv update into the peft module (see also the comment above)
```python
peft_block = self.peft_module(rank=rank, block=block)
self.peft_blocks.append(peft_block)
```

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,99 @@
+        self.w_a_linear_v = nn.Linear(self.dim, rank, bias=False)
+        self.w_b_linear_v = nn.Linear(rank, self.dim, bias=False)
+
+        self.reset_parameters()
```
Add the update to the block here. I.e.
```
blk.attn.qkv = self
```
This way the updates to the encoder are fully encapsulated here, and we have more flexibility and could e.g. later add a PEFT block that also updates the linear layers.

anwai98 replied:
Oh yes, this would have it completely independent. Thanks for the suggestion!

constantinpape reviewed (commented):
This is going the right direction! A few more changes to make it more modular.

In addition, it would also be good to have a test for this, that instantiates the PEFT SAM, applies a forward pass and checks the gradients afterwards.

anwai98 commented:
@constantinpape Thanks for the feedback. I've updated the `PEFT_Sam` block as mentioned, so that it is now independent of any assumptions to update parts of the backbone, and instead the `self.peft_module` will be responsible of doing this from now on (i.e. the surgery class provided by the user to handle everything for the attention blocks).

I also added a simple test for the `PEFT_Sam` class.

Let me know how things look now. Thanks!

anwai98 commented:
To add, looks like tests at `test_bioimageio/test_model_export.py` are failing for some reason.

constantinpape commented on micro_sam/training/peft_sam.py:
```diff
@@ -0,0 +1,103 @@
+                continue
+
+            peft_block = self.peft_module(rank=rank, block=blk)
+            self.peft_blocks.append(peft_block)
```
Can you do a
```python
self.peft_blocks = nn.ModuleList(self.peft_blocks)
```
after the loop? To make sure the gradients are propagated correctly.

constantinpape commented on test/test_peft_training.py:
```diff
@@ -0,0 +1,49 @@
+        # Check the expected shape of the outputs
+        mask_shapes = [output["masks"].shape[-2:] for output in outputs]
+        for shape in mask_shapes:
+            self.assertEqual(shape, input_shape[1:])
```
Ideally we would check here that gradients are propagated when through the model. But I can also do this in a follow up PR.

anwai98 replied:
Okay, I've left this in favor of the follow-up PR.

constantinpape reviewed (commented):
This looks good now, only one minor thing.

anwai98 commented:
Hi @constantinpape, this PR is GTG from my side now. Thanks!
